storm-core/src/jvm/org/apache/storm/daemon/supervisor/Slot.java

```diff
@@ -0,0 +1,776 @@
+        
+        if (dynamicState.newAssignment == null) {
+            return dynamicState.withState(MachineState.EMPTY);
+        }
+        Future<Void> pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment, staticState.port);
+        return dynamicState.withPendingLocalization(dynamicState.newAssignment, pendingDownload).withState(MachineState.WAITING_FOR_BASIC_LOCALIZATION);
+    }
+    
+    /**
+     * Kill the current container and start downloading what the new assignment needs, if there is a new assignment
+     * PRECONDITION: container != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception 
+     */
+    static DynamicState killContainerForChangedAssignment(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        
+        staticState.iSupervisor.killedWorker(staticState.port);
+        dynamicState.container.kill();
+        Future<Void> pendingDownload = null;
+        if (dynamicState.newAssignment != null) {
+            pendingDownload = staticState.localizer.requestDownloadBaseTopologyBlobs(dynamicState.newAssignment, staticState.port);
+        }
+        Time.sleep(staticState.killSleepMs);
+        return dynamicState.withPendingLocalization(dynamicState.newAssignment, pendingDownload).withState(MachineState.KILL);
+    }
+    
+    /**
+     * Kill the current container and relaunch it.  (Something odd happened)
+     * PRECONDITION: container != null
+     * @param dynamicState current state
+     * @param staticState static data
+     * @return the next state
+     * @throws Exception 
+     */
+    static DynamicState killAndRelaunchContainer(DynamicState dynamicState, StaticState staticState) throws Exception {
+        assert(dynamicState.container != null);
+        
+        dynamicState.container.kill();
+        Time.sleep(staticState.killSleepMs);
+        
+        //any stop profile actions that hadn't timed out yet, we should restart after the worker is running again.
+        HashSet<TopoProfileAction> mod = new HashSet<>(dynamicState.profileActions);
+        mod.addAll(dynamicState.pendingStopProfileActions);
+        return dynamicState.withState(MachineState.KILL_AND_RELAUNCH).withProfileActions(mod, Collections.<TopoProfileAction> emptySet());
+    }
+    
+    /**
+     * Clean up a container
+     * PRECONDITION: All of the processes have died.
+     * @param dynamicState current state
+     * @param staticState static data
+     * @param nextState the next MachineState to go to.
+     * @return the next state.
+     */
+    static DynamicState cleanupCurrentContainer(DynamicState dynamicState, StaticState staticState, MachineState nextState) throws Exception {
+        assert(dynamicState.container != null);
+        assert(dynamicState.currentAssignment != null);
+        
```

Comment:
That is the thing I really don't like putting anything into an assert with side effects, because it will behave differently between testing and production.  In this case it really should not matter, because the side effect is minimal, and it is just timing.